### DISCUSSION TITLE
Fastify support

### DISCUSSION THREAD

**Discussion by frbuceta** [nuxt/nuxt#17512] [Questions]
Hi everyone.

I have recently tried installing Express as an API on Nuxt. Shortly after I wanted to go a little further and try Fastify and here I ran into an error. My question is ... can it work?

**Express example (This works):**
```js
import express from 'express'
const app = express();
app.use(express.json());
app.get('/getJSON', (req, res) => {
  res.json({ data: 'data' })
})
export default app
```

**Fastify example (This does not work):**
```js
import fastify from 'fastify';

const app = fastify({logger: true})

app.route({
  method: 'GET',
  url: '/',
  schema: {
    // request needs to have a querystring with a `name` parameter
    querystring: {
      name: { type: 'string' }
    },
    // the response needs to be an object with an `hello` property of type 'string'
    response: {
      200: {
        type: 'object',
        properties: {
          hello: { type: 'string' }
        }
      }
    }
  },
  // this function is executed for every request before the handler is executed
  preHandler: async (request, reply) => {
    // E.g. check authentication
  },
  handler: async (request, reply) => {
    return { hello: 'world' }
  }
})

export default app
```

**Fastify error**
```
ERROR 500
handle is not a function
  at file://./node_modules/h3/dist/index.mjs:111:24
  at new Promise ()
  at callHandle (file://./node_modules/h3/dist/index.mjs:108:10)
  at file://./node_modules/h3/dist/index.mjs:104:12
  at file://./node_modules/h3/dist/index.mjs:132:34
  at processTicksAndRejections (node:internal/process/task_queues:96:5)
  at async handle (file://./node_modules/h3/dist/index.mjs:572:19)
```

Can someone give me a hand?

**Comment by wobsoriano** [comment]
Any reason why don't you just use the built-in [http framework](https://github.com/unjs/h3)?

**Reply by frbuceta** [reply]
Can you add a schema validator? It also doesn't have the GET, POST methods.

**Reply by wobsoriano** [reply]
You'll have to filter the request by `req.method`

**Reply by pi0** [reply]
Not direct answer to the discussion but h3 now has support for route matching and uses a super fast radix router :) As validator, I personally would use [joy](https://joi.dev/).

**Comment by wobsoriano** [comment]
Express works because it's close to h3's API. If you want, you can still use the built-in http framework and add in [yup](https://github.com/jquense/yup) for schema validation.

```ts
// server/api/post.ts
import type { IncomingMessage, ServerResponse } from 'http'
import { useQuery, useBody } from 'h3'
import * as yup from 'yup'

const schema = yup.object({
  body: yup.object({
    firstName: yup.string().required(),
    age: yup.number().required()
  }),
  query: yup.object({
    username: yup.string().required()
  }),
});

export default async (req: IncomingMessage, res: ServerResponse) => {
    if (req.method !== 'POST') {
        res.statusCode = 400
        res.end('Only POST requests allowed')
    }

    const query = useQuery(req)
    const body = useBody(req)

    try {
        await schema.validate({
            body,
            query
        });
    } catch (e) {
        res.statusCode = 400
        res.end('Invalid data')
    }

    return {
        success: true
    }
}
```

Alternative for yup: https://github.com/colinhacks/zod

**Reply by frbuceta** [reply]
The example is fine but I can't use parameters in the url with h3. For example: `/api/hello/:name` (I have already created an issue). With express this limitation I can skip it.

**Reply by wobsoriano** [reply]
Ah yes that's a limitation right now. 

https://github.com/nuxt/framework/discussions/1138

**Reply by wobsoriano** [reply]
> The example is fine but I can't use parameters in the url with h3. For example: `/api/hello/:name` (I have already created an issue). With express this limitation I can skip it.

We have this now

**Comment by calumk** [comment]
The same is true for other libraries - I am having issues with using [tinyhttp](https://github.com/talentlessguy/tinyhttp/)

The h3 implementation seems quite limiting

**Reply by wobsoriano** [reply]
With latest versions of Nuxt, I believe you cant just expose express instances or other http frameworks from middlewares right @pi0 ?

**Reply by pi0** [reply]
h3 is the core stack but you can always create a main API route exposing an standard `(req, res)` node handler from it from any framework. h3 has almost no interference. 

BTW what do you believe is limiting currently with h3 @calumk? Would be happy to discuss and improve :)
